albo.py

```diff
@@ -1,120 +1,63 @@
-        cached_announcements.append(id)
-    else:
-      if id in cached_announcements:
-        cached_announcements.remove(id)
-      send_telegram_message(message)
-      send_telegram_attachments(attachments)
-  except ValueError as err:
-    decoded_url = json.dumps(unquote(err.args[1]))
-    document_list = re.findall(r"https?:.+?(?=\\|\")", decoded_url)
-    error_string = "```response sent: " + err.args[0] + "\n```" + "related documents:\n" + escape_char('\n'.join(document_list))
-    print (json.dumps(unquote(err.args[1]))) #Insert here bot sending error message in group
-    for dev in CHAT_ID_DEV_LIST:
-      send_telegram_message(error_string, dev)
-  #print(message)
-
-with open("last_id.txt", "w+") as f_id:
-    f_id.write(str(new_id))
-
-with open("cached_announcements.txt", "w+") as f_cached:    
-  for cached_id in cached_announcements:
-      f_cached.write(str(cached_id) +  "\n")
+
+from utils.departments import find_all_departments
+from utils.formatting import escape_char, get_formatted_message
+from utils.resources import *
+from utils.tg import TelegramBot
+
+ALBO_UNICT_URL = "http://albo.unict.it/"
+
+tg_bot = TelegramBot.from_settings_file("settings.yaml")
+
+def main ():    
+    last_id = get_last_id("data/last_id.txt")
+
+    cached_announcements = get_cached_announcements("data/cached_announcements.txt")
+
+    page = requests.get(ALBO_UNICT_URL)
+
+    if page.status_code != 200:
+        print("Impossibile accedere al sito, riprovare più tardi")
+        quit(1)
+
+    soup = BeautifulSoup(page.content, 'html.parser')
+
+    table = soup.find('div', id='boge')
+
+    new_id = int(table.find('tr').find_next_sibling().td.string)
+
+    headers = [header.string for header in table.find('tr').find_all("td")]
+
+    ids_to_parse = cached_announcements + list(range(last_id + 1, new_id + 1))
+
+    for id in ids_to_parse:
+        tr = table.find('td', text=id).parent
+        row = tr.find_all('td')
+
+        message = get_formatted_message(row, headers)
+        tags = ' '.join(["*[" + department + "]*" for department in find_all_departments(message)])
+        message = tags + "\n\n" + message
+        
+        try:
+            attachments = [ALBO_UNICT_URL + list_item['href'] for list_item in tr.find_all('a')]
+            cached_announcements = update_cached_announcements(cached_announcements, id, len(attachments))
+
+            if len(attachments) != 0:
+                tg_bot.send_telegram_announcements(attachments, message)
+
+        except ValueError as err:
+            tg_bot.send_documents_error_message(err)
+
+    write_new_id("data/last_id.txt", new_id)
+    write_cached_announcements("data/cached_announcements.txt", cached_announcements)
+
+
+if __name__ == '__main__':
+    try:
+        main()
+    except Exception as e:
+        error_string = "An exception was raised:\n" + "`" + escape_char(traceback.format_exc()) +  "`"
+        tg_bot.send_debug_messages(error_string)
+        raise
```

Comment:
would be nice to say the kind of issue, like `raise error_string`, it does not cost much and who know, it can be helpful in the future beside the telegram message.

Reply:
The raise keyword without any argument just re-throws the precedently catched exception without modifying the stack trace, if a proper error message was set by the original exception it will not be overwritten

Reply:
> The raise keyword without any argument just re-throws the precedently catched exception without modifying the stack trace, if a proper error message was set by the original exception it will not be overwritten

Reference links:
https://stackoverflow.com/questions/36153805/difference-between-raise-and-raise-e/45986945#45986945
https://stackoverflow.com/questions/25001971/rethrowing-python-exception-which-to-catch